### ISSUE TITLE
0.13: [Bug]: JavaScript heap out of memory on `garden validate`

### ISSUE THREAD

**Issue by mdragilev** [garden-io/garden#4638]
### Current Behavior

After upgrading from 0.12.53 -> 0.13.[1|2], we are running into the an out of memory error.  Other commands like deploys run ok.  We use the validate command in our pull requests that impact garden resource definitions.

```
garden validate
Validate 

"apiVersion" is missing in the Project config. Assuming "garden.io/v0" for backwards compatibility with 0.12. The "apiVersion"-field is mandatory when using the new action Kind-configs. A detailed migration guide is available at https://docs.garden.io/tutorials/migrating-to-bonsai
Project configuration field `modules` is deprecated in 0.13 and will be removed in 0.14. Please use the `scan` field instead.
Garden v0.13 (Bonsai) is a major release with significant changes. Please help us improve it by reporting any issues/bugs here:
https://go.garden.io/report-bonsai
→ Run garden util hide-warning 0.13-bonsai to disable this warning.
ℹ garden               → Running in Garden environment thor.affirm-garden-affirm
ℹ providers            → Getting status...
ℹ providers            → Getting status...
✔ providers            → Done (took 0.2 sec)
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 0.2 sec)
ℹ providers            → Preparing environment...
ℹ local-kubernetes     → Configuring...
ℹ providers            → Getting status...
✔ providers            → Cached (took 0.2 sec)
ℹ providers            → Run with --force-refresh to force a refresh of provider statuses.
ℹ providers            → Getting status...
✔ providers            → Cached (took 0 sec)
ℹ providers            → Run with --force-refresh to force a refresh of provider statuses.
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 0.3 sec)
ℹ deploy.nginx-kind-new → outdated
ℹ deploy.nginx-kind-new → Deploying version v-7cc78df9b7...
ℹ deploy.nginx-kind-new → Waiting for resources to be ready...
ℹ deploy.nginx-kind-new → Resources ready
ℹ deploy.nginx-kind-new → Waiting for resources to be ready...
⚠ deploy.nginx-kind-new → Deployment/ingress-nginx-controller: MountVolume.SetUp failed for volume "webhook-cert" : secret "ingress-nginx-admission" not found
⚠ deploy.nginx-kind-new → Deployment/ingress-nginx-controller: RELOAD NGINX reload triggered due to a change in configuration
ℹ deploy.nginx-kind-new → Resources ready
✔ deploy.nginx-kind-new → Done (took 33.1 sec)
✔ local-kubernetes     → Ready
✔ providers            → Done (took 35.1 sec)
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 32.2 sec)

<--- Last few GCs --->

[389464:0x3f9e070]   200367 ms: Scavenge 4054.2 (4132.2) -> 4047.0 (4132.2) MB, 4.1 / 0.0 ms  (average mu = 0.583, current mu = 0.259) allocation failure; 
[389464:0x3f9e070]   200384 ms: Scavenge 4054.2 (4132.2) -> 4047.5 (4133.0) MB, 3.4 / 0.0 ms  (average mu = 0.583, current mu = 0.259) allocation failure; 
[389464:0x3f9e070]   200401 ms: Scavenge 4055.3 (4133.0) -> 4048.6 (4149.0) MB, 4.2 / 0.0 ms  (average mu = 0.583, current mu = 0.259) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0x9f2b20 node::Abort() [garden]
 2: 0x91b830 node::FatalError(char const*, char const*) [garden]
 3: 0x11c8d6e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [garden]
 4: 0x11c901a v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [garden]
 5: 0x13b43f5  [garden]
 6: 0x13c3dfd v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [garden]
 7: 0x139dc0d v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [garden]
 8: 0x139ed94 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [garden]
 9: 0x137e539 v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [garden]
10: 0x17618f6 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [garden]
11: 0x1ba0339  [garden]
Aborted (core dumped)
```

### Expected behavior

Garden validate runs

### Workaround

None?

### Suggested solution(s)

I do not have a suggested solution. I can generate debug or silly level logs, but I cannot post them in a public forum.  I can set up a dev environment for gardens and run any additional commands.

### Your environment

* **OS:** Ubuntu 22.04 or 18.04
* **How I'm running Kubernetes:** kind cluster on an EC2 node

0.13.2
0.13.1

**Comment by Orzelius** [comment]
@mdragilev Thank you for submitting this. Unfortunately there's little we can do without more information. Could you maybe provide us a higher level log of this via `garden validate -l5` and the output of `garden get config`.
Perhaps it would be possible to add them to a private gist and give @garden-io/core members access?

If you can't go with that option maybe you could do some digging on your own.
The silly log level logs should include lines like `Calling validate handler for action <action reference>`
Probably the issue lies within the last few actions validated.

**Comment by mdragilev** [comment]
Hi.  For the `garden get config` I also get an OOM error, so for `garden validate` it never seems to get to `Calling validate...`

```
affirm@mdragilev-thor-v2-tauros:~/all-the-things$ garden get config
Get config

"apiVersion" is missing in the Project config. Assuming "garden.io/v0" for backwards compatibility with 0.12. The "apiVersion"-field is mandatory when using the new action Kind-configs. A detailed migration guide is available at https://docs.garden.io/tutorials/migrating-to-bonsai
Project configuration field `modules` is deprecated in 0.13 and will be removed in 0.14. Please use the `scan` field instead.
Garden v0.13 (Bonsai) is a major release with significant changes. Please help us improve it by reporting any issues/bugs here:
https://go.garden.io/report-bonsai
→ Run garden util hide-warning 0.13-bonsai to disable this warning.
ℹ garden               → Running in Garden environment thor.affirm-garden-affirm
ℹ providers            → Getting status...
✔ providers            → Cached (took 0.6 sec)
ℹ providers            → Run with --force-refresh to force a refresh of provider statuses.
ℹ graph                → Resolving actions and modules...
✔ graph                → Done (took 32.6 sec)

<--- Last few GCs --->

[2508509:0x44d90c0]   159263 ms: Mark-sweep 4030.3 (4141.8) -> 4015.1 (4142.8) MB, 1331.6 / 0.0 ms  (average mu = 0.278, current mu = 0.023) allocation failure; scavenge might not succeed
[2508509:0x44d90c0]   161258 ms: Mark-sweep 4031.4 (4143.1) -> 4016.2 (4144.1) MB, 1965.6 / 0.0 ms  (average mu = 0.141, current mu = 0.014) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0x9f2b20 node::Abort() [garden]
 2: 0x91b830 node::FatalError(char const*, char const*) [garden]
 3: 0x11c8d6e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [garden]
 4: 0x11c901a v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [garden]
 5: 0x13b43f5  [garden]
 6: 0x13c3dfd v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [garden]
 7: 0x139dc0d v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [garden]
 8: 0x139ed94 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [garden]
 9: 0x137e539 v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [garden]
10: 0x17618f6 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [garden]
11: 0x1ba0339  [garden]
Aborted (core dumped)
```

I can't share private gists.  I did create a repository with logs that I can share.  I cannot add @garden-io/core as that seems to be a private team.  Is there someone besides yourself I need to share with?

**Comment by vvagaytsev** [comment]
Thanks, @mdragilev! Can you share the repo with me, please?

**Comment by mdragilev** [comment]
@vvagaytsev I sent an invite to you for the https://github.com/mdragilev/share_with_garden repository.  Sorry for the delay, was on vacation.

**Comment by vvagaytsev** [comment]
@mdragilev thank you so much!

**Comment by vvagaytsev** [comment]
@mdragilev it looks like the https://github.com/mdragilev/share_with_garden/blob/main/garden_validate_l5.log file is truncated, there is nothing about OOM.

Could you rerun it with `-l4` and share the complete file, please?

**Comment by mdragilev** [comment]
@vvagaytsev I re-ran with `-l4` and `-l5` flags.  I needed to redirect the stderr to capture the OOM at the end.  The `-l5` logs are zipped as they were >300MB otherwise. Thank you again for taking a look.

**Comment by mdragilev** [comment]
Same results with the newly released CLI.  Do we open up a new issue?

**Comment by vvagaytsev** [comment]
Thanks for the update @mdragilev! I'm re-opening this one.

**Comment by mdragilev** [comment]
@vvagaytsev Got it.  I updated https://github.com/mdragilev/share_with_garden for 0.13.11 version of the CLI.

**Comment by shumailxyz** [comment]
@mdragilev A fix for the memory leak has been released in 0.13.15. Please give it a spin and see if it resolves the out of memory issue for you.

**Comment by mdragilev** [comment]
Thank you.  This has addressed the bug.
